pkg/snapshot/snapshotter/snapshotter_test.go

```diff
@@ -452,6 +452,386 @@ var _ = Describe("Snapshotter", func() {
+				}
+				fmt.Println("Current hour full snapshot list prepared")
+
+				// delta snapshots
+				snapTime = snapTime.Add(time.Duration(-time.Minute * 30))
+				snapTime = snapTime.Add(time.Duration(time.Minute * 10))
+				for now.Sub(snapTime) >= 0 {
+					snap := &brtypes.Snapshot{
+						Kind:          brtypes.SnapshotKindDelta,
+						CreatedOn:     snapTime,
+						StartRevision: 0,
+						LastRevision:  1001,
+					}
+					snap.GenerateSnapshotName()
+					expectedSnapList = append(expectedSnapList, snap)
+					snapTime = snapTime.Add(time.Duration(time.Minute * 10))
+				}
+				fmt.Println("Incremental snapshot list prepared")
+
+				//start test
+				snapshotterConfig := &brtypes.SnapshotterConfig{
+					FullSnapshotSchedule:     schedule,
+					DeltaSnapshotPeriod:      wrappers.Duration{Duration: 10 * time.Second},
+					DeltaSnapshotMemoryLimit: brtypes.DefaultDeltaSnapMemoryLimit,
+					GarbageCollectionPeriod:  wrappers.Duration{Duration: garbageCollectionPeriod},
+					GarbageCollectionPolicy:  brtypes.GarbageCollectionPolicyExponential,
+					MaxBackups:               maxBackups,
+				}
+
+				ssr, err := NewSnapshotter(logger, snapshotterConfig, store, etcdConnectionConfig, compressionConfig)
+				Expect(err).ShouldNot(HaveOccurred())
+
+				gcCtx, cancel := context.WithTimeout(testCtx, testTimeout)
+				defer cancel()
+				ssr.RunGarbageCollector(gcCtx.Done())
+
+				list, err := store.List()
+				Expect(err).ShouldNot(HaveOccurred())
+				Expect(len(list)).Should(Equal(len(expectedSnapList)))
+
+				for index, snap := range list {
+					if snap.CreatedOn != expectedSnapList[index].CreatedOn || snap.Kind != expectedSnapList[index].Kind {
+						Fail("Expected snap list doesn't match with output snap list")
+					}
+				}
+			})
+
+			//Test to check backward compatibility of garbage collector
+			//Tests garbage collector behaviour (in exponential config) when only v1 directory is present
+			//TODO: Consider removing when backward compatibility no longer needed
+			It("should garbage collect exponentially with only v1 dir structure present (backward compatible test)", func() {
+				logger.Infoln("creating expected output")
+
+				// Prepare expected resultant snapshot list
+				var (
+					now = time.Now().UTC()
+					//store            = prepareBackwardCompatibleStoreInV1ForGC(now, "gc_exponential_backward_compatiblev1.bkp")
+					store            = prepareStoreForGarbageCollectionInPrefix(now, "gc_exponential_backward_compatiblev1.bkp", "v1")
+					snapTime         = time.Date(now.Year(), now.Month(), now.Day()-35, 0, -30, 0, 0, now.Location())
+					expectedSnapList = brtypes.SnapList{}
+				)
+
+				// weekly snapshot
+				for i := 1; i <= 4; i++ {
+					snapTime = snapTime.Add(time.Duration(time.Hour * 24 * 7))
+					snap := &brtypes.Snapshot{
+						Kind:          brtypes.SnapshotKindFull,
+						CreatedOn:     snapTime,
+						StartRevision: 0,
+						LastRevision:  1001,
+					}
```

Comment:
After the garbage collection, is the whole store expected to be in the `v2` format? This is fine, I am just asking because I did not find `SnapDir` field being initialised for any of the  expected snapshots.

Also, is there a commonality in the initialisation of the expected snapshot lists between the cases? If so, can you please convert the logic into a function and reuse?

Reply:
No, not really. Snaps that are not garbage collected would remain in their original directory.
Added changes to support `SnapDir` in the expected snapshots and using that to decide test case result